### PR TITLE
Bech32 address improvements

### PR DESCRIPTION
**Pull request by benthecarman**
Fixes so we can read uppercase addresses, adds some decoding optimizations, and makes sure we verify the checksum of the address.

### CODE REVIEW

File: core/src/main/scala/org/bitcoins/core/protocol/Address.scala [hunk 1]
```diff
@@ -113,7 +117,7 @@ object Bech32Address extends AddressFactory[Bech32Address] {
       networkParameters: NetworkParameters,
       data: Vector[UInt5])
       extends Bech32Address {
-    //require(verifyChecksum(hrp, data), "checksum did not pass")
+    require(verifyChecksum, "checksum did not pass")
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/protocol/Address.scala]: nit: Put them in the error messages

Comment (Christewart) [reply core/src/main/scala/org/bitcoins/core/protocol/Address.scala]: ooh wow, this isn't applied to _any_ of our address types. We probably want to add this in the super class possibly? 

I'm going to file an issue

https://github.com/bitcoin-s/bitcoin-s/blob/223be8057351017eb0d454b97b36a917f00fe1ac/core/src/main/scala/org/bitcoins/core/protocol/Address.scala#L215

File: core/src/main/scala/org/bitcoins/core/util/Bech32.scala [hunk 2]
```diff
@@ -305,6 +305,17 @@ object Bech32 extends Bech32 {
   val charset: Vector[Char] = Vector('q', 'p', 'z', 'r', 'y', '9', 'x', '8',
     'g', 'f', '2', 't', 'v', 'd', 'w', '0', 's', '3', 'j', 'n', '5', '4', 'k',
     'h', 'c', 'e', '6', 'm', 'u', 'a', '7', 'l')
+
+  /** The Bech32 character set for decoding. */
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/util/Bech32.scala]: is there a reference to this? Is it reversed or the inverse?

**Review by Christewart — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/bitcoins/core/util/Bech32.scala (hunk 2 above)
Comment (benthecarman) [reply core/src/main/scala/org/bitcoins/core/util/Bech32.scala]: Added link from where I got it from

**Review by Christewart — APPROVED** [review]

**Pull request merged**